remove/remove.go

```diff
@@ -0,0 +1,199 @@
+	Variables struct {
+		Name    string `json:"name"`
+		EnvPath string `json:"envPath"`
+	} `json:"variables"`
+}
+
+func Remove(conf *config.Config, s S3Remover, envPath, version string) error {
+	envPath = filepath.Clean(string(filepath.Separator) + envPath)[1:]
+
+	modulePath := filepath.Join(conf.Module.ModuleInstallDir, envPath)
+	scriptPath := filepath.Join(conf.Module.ScriptsInstallDir, envPath, version+build.ScriptsDirSuffix)
+
+	if err := checkWriteAccess(modulePath, scriptPath); err != nil {
+		return err
+	}
+
+	if err := removeEnvFromCore(conf, envPath+"-"+version); err != nil {
+		return err
+	}
+
+	if err := removeLocalFiles(modulePath, scriptPath); err != nil {
+		return err
+	}
+
+	return removeFromS3(s, modulePath)
+}
+
+func checkWriteAccess(modulePath, scriptPath string) error {
+	for _, p := range [...]string{
+		filepath.Dir(modulePath),
+		modulePath,
+		filepath.Dir(filepath.Dir(scriptPath)),
+		filepath.Dir(scriptPath),
+		scriptPath,
+	} {
+		if err := unix.Access(p, unix.W_OK); err != nil {
+			return Error(fmt.Sprintf("no write access to dir (%s): %s", p, err))
+		}
+	}
+
+	return nil
+}
+
+func removeEnvFromCore(conf *config.Config, envPath string) error {
+	log.Printf("Removing env %s from core\n", envPath)
+
+	req, err := http.NewRequestWithContext(
+		context.Background(),
+		http.MethodPost,
+		strings.TrimSuffix(conf.CoreURL, "/")+graphQLEndpoint,
+		createGraphQLPacket(envPath),
+	)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	return handleCoreResponse(resp)
+}
+
+func createGraphQLPacket(envPath string) io.Reader {
+	mutation := graphQLDeleteEnvironmentMutation{
+		Query: graphQLDeleteEnvironment,
+	}
+	mutation.Variables.Name = filepath.Base(envPath)
+	mutation.Variables.EnvPath = filepath.Dir(envPath)
+
+	var buf bytes.Buffer
+
+	json.NewEncoder(&buf).Encode(mutation) //nolint:errcheck
+
+	return &buf
+}
+
+func handleCoreResponse(resp *http.Response) error {
+	var cr coreResponse
+
+	err := json.NewDecoder(resp.Body).Decode(&cr)
+	if err != nil {
+		return err
+	}
+
+	if cr.Data.DeleteEnvironment.Name != "" {
+		return Error(cr.Data.DeleteEnvironment.Message)
+	}
+
+	return nil
+}
+
+func removeLocalFiles(modulePath, scriptPath string) error {
+	if err := removeAllNoDescend(modulePath); err != nil {
+		return err
+	}
+
+	return removeAllNoDescend(scriptPath)
+}
+
+func removeAllNoDescend(path string) error {
+	files, err := os.ReadDir(path)
+	if err != nil {
+		return err
+	}
+
+	for _, file := range files {
+		toRemove := filepath.Join(path, file.Name())
+
+		log.Printf("Removing file: %s\n", toRemove)
+
+		if err := os.Remove(toRemove); err != nil {
+			return err
+		}
+	}
+
+	log.Printf("Removing directory: %s\n", path)
+
+	return os.Remove(path)
+}
+
+var files = [...]string{"build.out", "singularity.def", "singularity.sif", "executables"} //nolint:gochecknoglobals
```

Comment:
Move to under import and call it something like s3BasenamesForDeletion. These basenames ought to also be consts from the package where we upload these files. Bonus points if there's a reasonable way of having that package tell us these basenames.